### ISSUE TITLE
[no-unnecessary-type-assertion] False positives with ! non-null assertion

### ISSUE THREAD

**Issue by bisubus** [typescript-eslint/typescript-eslint#2180]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

```TS
class Foo {
  protected _bar?: Bar;
  
  bar() {
    if (!this._bar) {
      this._bar = new Bar();
    }
    
    return this._bar!.baz();
  }
}
```

**Expected Result**

Should not cause an error because the assertion is not a no-op.

**Actual Result**

Causes an error:

> (property) Foo ._bar?: Bar
>
> This assertion is unnecessary since it does not change the type of the expression.eslint@typescript-eslint/no-unnecessary-type-assertion

**Additional Info**

TS config has strict mode enabled, `strict: true`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.31.0` |
| `@typescript-eslint/parser`        | `2.31.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `14.2.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
This is working as intended - you do not need the `!`. 
[repl](https://www.typescriptlang.org/play/#code/MYGwhgzhAEBCYCdoG9oCMwC8AUBKF0AvkQFAmiQwBiA9jSidNAA4I0AuApsFwCbQB9DAgD8ALjiIA3GSbC8DJkwCWAM2jYAhOwAWyiADohifMkZKmu-UeHQAvNAB2nAO6SEeGRcKylCTuwArgiO0FaGxggGGDi4XqSEQA)
